resolves #1880 allow theme to customize quotation mark characters

klonfish opened the pull request:
This adds the possibility to set the characters used as quotation marks via attributes. That way, they can be adapted to the style used in different languages.

I'm not quite happy with using four new attributes for this small feature, but I also couldn't find a better alternative. One alternative would be to use only one attribute per type (double and single) to define the opening and closing marks at once, but then they would have to be separated somehow and it would require some error handling (e.g. to catch if the user specified only one or more than two marks).

Additionally, the new attributes should probably be mentioned somewhere in the docs. Is there already a good place or should that go to a whole new section in the README?

Initially, I tried to read the attributes only once in `init_pdf` and store them in an instance variable to be used in `convert_inline_quoted`. However, this lead to a crash if smart quotes where used in a section title *and* that section did not have a custom ID assigned (or `sectids` was disabled). I suspect that the auto-generation of the section IDs calls `convert` on the title nodes before `init_pdf` gets called. Therefore I had to do the lookup in `convert_inline_quoted` and hope that this doesn't impact performance too much.

mojavelinux commented:
I like this idea. However, I think it should be controlled by the theme rather than the document. To control it via the document would mean that we need to introduce it into the Asciidoctor core, and like the language itself. The theme gives us a way to make customizations sooner, in a private space.

klonfish commented:
Thanks for you feedback. I also considered putting the values into the theme, but I think such a basic functionality shouldn't be the reason to make them language dependent. Otherwise, if one wants to convert documents with different languages (which is my usecase), there would have to be a different theme for all of them. Since the document itself obviously already is language-dependent, I went with the current implementation.

Also, I don't fully understand why it would have to be handled by the core. I took a quick look into how the quotation marks are handled by it and that seems to be quite language-agnostic. Especially the fact, that they are converted to `<quote>` tags in DocBook and that [their documentation](https://tdg.docbook.org/tdg/5.0/quote.html) explicitly states that they should be handled according to the `lang` attribute, led me to the assumption that they should be handled when generating the final output.
Additionally, isn't the `:hyphen:` attribute a very similar case, which is also handled only in `asciidoctor-pdf` and can even have a language different to the `lang` attribute of the core?

mojavelinux commented:
> Additionally, isn't the :hyphen: attribute a very similar case, which is also handled only in asciidoctor-pdf and can even have a language different to the lang attribute of the core?

You correctly point out that the hyphenation feature should probably be controlled by the theme.

mojavelinux commented:
If we do decide to allow the smart quotation characters to be controlled via document attributes, this is something that needs to be addressed first in core, and perhaps even in the AsciiDoc spec. By using the theme, we can advance this feature in Asciidoctor PDF (as it does not impact AsciiDoc language compliance).

klonfish commented:
Alright then, I updated the PR such that it uses the theme to define the quotation mark characters and also added some documentation for it in the theming guide.
Is there any ordering of the key categories in theming guide? Since I couldn't figure out any, I put it where I think it would fit best.

Do you think it would make sense to add `-content` to the keys? In the current state it would only make the keys longer, but it would allow for future extensions, e.g. someone might want to add a way to style the quotation marks differently. Then again, this might be too specific and out of scope for this project.

mojavelinux commented:
I like this proposal. For the theme structure, I would prefer to try to emulate CSS closer since that is the inspiration for the theme's schema. That would look like:

```
quotes: [«, », ‹, ›]
```

I'm unsure whether to organize it under the "base" category, but that might just be unnecessary.

I'll go ahead and apply these revisions as a proposal on your PR so you can see what I have in mind. Then we can decide how to proceed.

mojavelinux commented:
I have updated the PR. Please review when you get a chance.

mojavelinux commented:
> Is there any ordering of the key categories in theming guide?

I try to order it based on related things. Since this is a top-level entry, I think it belongs just under the Vertical Space, which is where I moved it to.

klonfish commented:
Thanks for your suggestion. I like the new format, as it less bloated than my version. It didn't occur to me that just using an array is an option that is clean and still allows for multiple characters for each type.

However, I find it a bit counterintuitive, that the double quotes are also used as single quotes, if only two are specified, but it seems that CSS behaves the same way. And although the concept of nested quote blocks in HTML is a bit different from the explicit usage of ``"` `"`` and ``'` `'`` in AsciiDoc, it's probably still not a big deal, as I guess that most users will either set all four types or leave them all unchanged.

Is there an easy way to issue warnings during compilation? It'd probably be nice to inform the user that the output might not be as expected, if they did not specify either two or four elements or if it isn't an array at all.

Additionally, I thought of two further test cases: multiple characters per type (e.g. French has spaces inside of the quotes) and the usage of the double quotes as single quotes, if only two elements are given (if we keep that). I can add those, if you also think that they are relevant and that they won't be too much.

> I'm unsure whether to organize it under the "base" category, but that might just be unnecessary.

I probably don't yet have enough insight into this project to give a good opinion on that, but as far as I see, the `base` keys can all be overwritten in other places of the theme. So, if that's a criterion, the quotation marks won't fit there.

mojavelinux commented on lib/asciidoctor/pdf/converter.rb:
```diff
@@ -472,10 +472,12 @@ def prepare_theme theme
-        theme.quotation_mark_single_close ||= '&#8217;'
+        if ::Array === (quotes = theme.quotes)
+          quotes[2] ||= quotes[0]
+          quotes[1] ||= quotes[1]
```
Uh oh. You're right!

mojavelinux commented:
> However, I find it a bit counterintuitive, that the double quotes are also used as single quotes, if only two are specified, but it seems that CSS behaves the same way. And although the concept of nested quote blocks in HTML is a bit different from the explicit usage of "` `" and '` `' in AsciiDoc, it's probably still not a big deal, as I guess that most users will either set all four types or leave them all unchanged.

I do agree this behavior in CSS is strange. Perhaps we could slightly modify by saying that if you only specify two, then we use our own default for single quotes. Likely, the user is going to specify all 4, but still.

> Is there an easy way to issue warnings during compilation? It'd probably be nice to inform the user that the output might not be as expected, if they did not specify either two or four elements or if it isn't an array at all.

See #1876. It's not something I'm going to work on, but if someone is so inclined...

> multiple characters per type

I don't see anything preventing this from working. We are accepting a character sequence, not a single character.

> the base keys can all be overwritten in other places of the theme

I was leaving room for the possibility that, one day, quotes could be customized per block. But, like you said, I'm not really sure what the right call is here. We can always move it later if we have reason to, I suppose.
